source/common/upstream/health_checker_impl.cc

```diff
@@ -405,6 +405,19 @@ void TcpHealthCheckerImpl::TcpActiveHealthCheckSession::onEvent(uint32_t events)
       events & Network::ConnectionEvent::LocalClose) {
     parent_.dispatcher_.deferredDelete(std::move(client_));
   }
+
+  if ((events & Network::ConnectionEvent::Connected) && parent_.receive_bytes_.empty()) {
+    // In this case we are just testing that we can connect, so immediately succeed. Also, since
+    // we are just doing a connection test, close the connection.
+    // NOTE(mattklein123): I've seen cases where the kernel will report a successful connection, and
+    // then proceed to fail subsequent calls (so the connection did not actually succeed). I'm not
+    // sure what situations cause this. If this turns into a problem, we may need to introduce a
+    // timer and see if the connection stays alive for some period of time while waiting to read.
+    // (Though we may never get a FIN and won't know until if/when we try to write). In short, this
+    // may need get more complicated but we can start here.
```

Comment:
What if you try to read from the socket? It should return `EAGAIN` in case there is nothing to read (success) or `0` if connection was closed.